src/decoder/lattice-faster-decoder-combine-bucketqueue.cc

```diff
@@ -0,0 +1,1159 @@
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#include "decoder/lattice-faster-decoder-combine.h"
+#include "lat/lattice-functions.h"
+
+namespace kaldi {
+
+template<typename Token>
+BucketQueue<Token>::BucketQueue(BaseFloat best_cost_estimate,
+                                BaseFloat cost_scale) :
+    cost_scale_(cost_scale) {
+  // NOTE: we reserve plenty of elements to avoid expensive reallocations
+  // later on. Normally, the size is a little bigger than (adaptive_beam +
+  // 5) * cost_scale.
+  int32 bucket_size = 100;
+  buckets_.resize(bucket_size);
+  bucket_storage_begin_ = std::floor((best_cost_estimate - 15) * cost_scale);
+  first_occupied_bucket_index_ = bucket_storage_begin_ + bucket_size;
+}
+
+template<typename Token>
+void BucketQueue<Token>::Push(Token *tok) {
+  int32 bucket_index = std::floor(tok->tot_cost * cost_scale_);
+  size_t vec_index = static_cast<size_t>(bucket_index - bucket_storage_begin_);
+
+  if (vec_index >= buckets_.size()) {
+    KALDI_WARN << "Have to reallocate the BucketQueue. Maybe need to reserve"
+               << " more elements in constructor.";
+    int32 offset = static_cast<int32>(vec_index);
+    // a margin here (e.g. 10);
+    int32 increase_size = offset >= 0 ? offset + 1 - buckets_.size() + 10 :
+                                        - offset + 10;
+    buckets_.resize(buckets_.size() + increase_size);
+    
+    // Push front
+    if (offset < 0) {
+      std::vector<std::vector<Token*> > tmp(buckets_);
+      buckets_.clear();
+      for (int32 i = 10 - offset ; i < buckets_.size(); i++) {
+        buckets_[i].swap(tmp[i + offset - 10]);
+      }
+      // Update start point
+      bucket_storage_begin_ = bucket_index - 10;
+      vec_index = 10;
+    }
+  }
+
+  tok->in_queue = true;
+  buckets_[vec_index].push_back(tok);
+  if (vec_index < (first_occupied_bucket_index_ - bucket_storage_begin_))
+    first_occupied_bucket_index_ = vec_index + bucket_storage_begin_;
+}
+
+template<typename Token>
+Token* BucketQueue<Token>::Pop() {
+  int32 vec_index = first_occupied_bucket_index_ - bucket_storage_begin_;
+  Token* best_tok = NULL;
+  while(vec_index < buckets_.size()) {
+    // Remove the best token
+    best_tok = buckets_[vec_index].back();
+    buckets_[vec_index].pop_back();
+
+    if (buckets_[vec_index].empty()) {  // This bucket is empty. Update
+                                        // first_occupied_bucket_index_
+      int32 next_vec_index = vec_index + 1;
+      for(; next_vec_index < buckets_.size(); next_vec_index++) {
+        if(!buckets_[next_vec_index].empty()) break;
+      }
+      first_occupied_bucket_index_ = bucket_storage_begin_ + next_vec_index;
+      vec_index = next_vec_index;
+    }
+
+    if (best_tok->in_queue) {  // This is a effective token
+      best_tok->in_queue = false;
```

Comment:
I would make 'best_tok' just a variable inside the loop called 'took', and at this point, instead of doing `break;`, just return.  And no 'else' statement.  Then below, instead of 'return best_tok;', just do 'return NULL; '